### ISSUE TITLE
Add comparison of error quality

### ISSUE THREAD

**Issue by epage** [epage/parse-benchmarks-rs#4]
The main question is how we should visualize this for the user

**Comment by jsinger67** [comment]
@0x2a-42,
first let me say that it's a great idea to compare error handling.
Parol provides a mean to hook into the error handling by implementing the Report trait and what you see here is the admittedly very verbose default implementation. When starting a new parser project a user is possibly grateful for having detailed error information. When a project gets more mature the error reporting can adjusted as needed.

**Comment by zesterer** [comment]
@0x2a-42 Thanks for giving all of those a go! This looks like a fun comparison.

As a potentially interesting data point, here's the error when using the JSON example in chumsky (i.e: prioritising error quality over speed): 

![Image](https://github.com/user-attachments/assets/ff2777f8-ae58-45fe-bf8c-dc8f3e2fc312)

I think as mentioned, it's going to be very difficult to usefully compare both error quality and performance using the same implementations.

**Comment by 0x2a-42** [comment]
@ltratt: Your implementation seems to panic during an `unwrap` in a semantic action of the parser.
```rust
let s = $lexer.span_str($1.unwrap().span());
```
It would be unfortunate if your code could not be used, as you mentioned in 631687e, that it supports error recovery.

@jsinger67: OK, that makes sense. I think it would be more convenient for users, if they could restrict the output to syntax errors, without having to implement the `Report` trait themselves, as the default implementation seems to require quite a lot of code.

@zesterer:
> As a potentially interesting data point, here's the error when using the JSON example in chumsky (i.e: prioritising error quality over speed):

Interesting, so when prioritizing error quality, the parser still stops after the first error. Maybe this example could help to improve the error recovery in a future version of chumsky. If the missing colon is added, the other two errors are however reported correctly.

> I think as mentioned, it's going to be very difficult to usefully compare both error quality and performance using the same implementations.

I think a fair comparison would require a single implementation for each library. It is however also true, that not every use case requires good error messages, so a separate implementation would also provide valuable information. It would probably make sense, to add an optional time measurement for libraries that allow to optimize the parser at the cost of the error quality.

Name | Overhead (release) | Build (debug) | Parse (release) | Downloads | Version
-----|--------------------|---------------|-----------------|-----------|--------
null | 0 KiB | 217ms | 4ms | - | -
chumsky | 117 KiB | 4s | 85ms (31ms [^1]) | ![Download count](https://img.shields.io/crates/dr/chumsky) | v0.10.1

[^1]: Without error recovery.

**Comment by zesterer** [comment]
> Interesting, so when prioritizing error quality, the parser still stops after the first error. Maybe this example could help to improve the error recovery in a future version of chumsky. If the missing colon is added, the other two errors are however reported correctly.

Recovery is an as-it-comes affair, and you're at liberty to add your own recovery strategies. If I add

```rs
.recover_with(via_parser(value))
```

to the `object` parser, I get the following:

![Image](https://github.com/user-attachments/assets/f156227f-0076-4c37-8090-876479d33ae3)

I think this just goes to show how difficult it is to really pull out any particular conclusions about error quality from a small example.

**Comment by 0x2a-42** [comment]
@zesterer:
> Recovery is an as-it-comes affair, and you're at liberty to add your own recovery strategies.

That is true for libraries where error recovery involves a manual process. Error recovery in [LALRPOP](https://lalrpop.github.io/lalrpop/tutorial/008_error_recovery.html) also seems to involve annotating the grammar, similar to how error rules were used in Bison. Other libraries use heuristics to automate this process, so there is value in comparing how well those handle realistic cases. But even for manual error recovery you can get a feeling for how much it slows down and complicates the parser.

> If I add
> ```rust
> .recover_with(via_parser(value))
> ```
> to the object parser, I get the following three errors all at once:

Nice! Could you be more precise about the point at which this call has to be inserted? I was trying to reproduce this but couldn't get it to work.

> I think this just goes to show how difficult it is to really pull out any particular conclusions about error quality from a small example.

It is certainly difficult to exactly quantify how well a parser handles errors, as this is to some degree subjective. However it is possible to give potential users an impression of what is possible with a given library.

Yes, one example would probably not suffice. It would be better to use a more diverse set of examples, where the intent would be obvious to most humans, so a correct number of errors can be defined. But even with a single example you can already see quite a bit of differences between the available parsers. An advantage of a simple reference example could also be, that it is short enough to allow potential users to quickly judge the quality of the errors themselves.

@ltratt: Thanks, using `?` instead of `unwrap` indeed fixes the problem. Afterwards the parser reports two of the three errors.
```
Parsing error at line 1 column 7. Repair sequences found:
   1: Insert :, Shift [, Delete 1
   2: Insert :, Shift [, Shift 1, Delete 2
   3: Insert :, Shift [, Shift 1, Insert ,
Parsing error at line 1 column 27. Repair sequences found:
   1: Insert STRING, Insert :, Insert TRUE
   2: Insert STRING, Insert :, Insert NULL
   3: Insert STRING, Insert :, Insert FLOAT
   4: Insert STRING, Insert :, Insert FALSE
   5: Insert STRING, Insert :, Insert STRING
```

**Comment by zesterer** [comment]
> Nice! Could you be more precise about the point at which this call has to be inserted? I was trying to reproduce this but couldn't get it to work.

I told a very minor white lie: I did end up slightly restructuring the `member` parser (effectively just inverting the nested `then`s, which is a change that has no effect on the parser's ordinary behaviour). Still, it's a trivial edit.

```diff
diff --git a/examples/json.rs b/examples/json.rs
index f952a38..914fe82 100644
--- a/examples/json.rs
+++ b/examples/json.rs
@@ -88,7 +88,8 @@ fn parser<'a>() -> impl Parser<'a, &'a str, Json, extra::Err<Rich<'a, char>>> {
             )
             .boxed();

-        let member = string.clone().then_ignore(just(':').padded()).then(value);
+        let member = string.clone().then(just(':').padded().ignore_then(value.clone())
+            .recover_with(via_parser(value)));
         let object = member
             .clone()
             .separated_by(just(',').padded().recover_with(skip_then_retry_until(
```
